File: FWFramework/Extend/UIKit/UIScreen+FWFramework.m
//
//  UIScreen+FWFramework.m
//  FWFramework
//
//  Created by wuyong on 17/3/13.
//  Copyright © 2017年 ocphp.com. All rights reserved.
//

#import "UIScreen+FWFramework.h"

@implementation UIScreen (FWFramework)

+ (CGSize)fwScreenSize
{
    return [UIScreen mainScreen].bounds.size;
}

+ (CGFloat)fwScreenWidth
{
    return [UIScreen mainScreen].bounds.size.width;
}

+ (CGFloat)fwScreenHeight
{
    return [UIScreen mainScreen].bounds.size.height;
}

+ (CGFloat)fwScreenScale
{
    return [UIScreen mainScreen].scale;
}

+ (CGSize)fwScreenResolution
{
    return CGSizeMake(
                      [UIScreen mainScreen].bounds.size.width * [UIScreen mainScreen].scale,
                      [UIScreen mainScreen].bounds.size.height * [UIScreen mainScreen].scale
                      );
}

+ (BOOL)fwIsScreenSize:(CGSize)size
{
    return CGSizeEqualToSize(size, [UIScreen mainScreen].bounds.size);
}

+ (BOOL)fwIsScreenResolution:(CGSize)resolution
{
    return CGSizeEqualToSize(resolution, [self fwScreenResolution]);
}

+ (BOOL)fwIsScreen35
{
    return [self fwIsScreenSize:CGSizeMake(320, 480)];
}

+ (BOOL)fwIsScreen40
{
    return [self fwIsScreenSize:CGSizeMake(320, 568)];
}

+ (BOOL)fwIsScreen47
{
    return [self fwIsScreenSize:CGSizeMake(375, 667)];
}

+ (BOOL)fwIsScreen55
{
    return [self fwIsScreenSize:CGSizeMake(414, 736)];
}

+ (BOOL)fwIsScreen58
{
    return [self fwIsScreenSize:CGSizeMake(375, 812)];
}

+ (BOOL)fwIsScreen61
{
    return [self fwIsScreenResolution:CGSizeMake(828, 1792)];
}

+ (BOOL)fwIsScreen65
{
    return [self fwIsScreenResolution:CGSizeMake(1242, 2688)];
}

+ (BOOL)fwIsScreenX
{
    return [self fwIsScreenSize:CGSizeMake(375, 812)] || [self fwIsScreenSize:CGSizeMake(414, 896)];
}

+ (CGFloat)fwStatusBarHeight
{
    return [self fwIsScreenX] ? 44.0 : 20.0;
}

+ (CGFloat)fwNavigationBarHeight
{
    return [self fwIsScreenX] ? 44.0 : 44.0;
}

+ (CGFloat)fwTabBarHeight
{
    return [self fwIsScreenX] ? 83.0 : 49.0;
}

@end

@implementation UIViewController (FWScreen)

+ (BOOL)fwIsControllerStatusBar
{
    static BOOL isController;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        // 读取当前UIViewControllerBasedStatusBarAppearance设置，默认YES视图控制器生效，NO则UIApplication生效
        id object = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIViewControllerBasedStatusBarAppearance"];
        isController = (object && ![object boolValue]) ? NO : YES;
    });
    return isController;
}

- (CGFloat)fwStatusBarHeight
{
    if ([self.class fwIsControllerStatusBar]) {
        if ([self prefersStatusBarHidden]) {
            return 0.0;
        } else {
            return [UIApplication sharedApplication].statusBarFrame.size.height;
        }
    } else {
        if ([UIApplication sharedApplication].statusBarHidden) {
            return 0.0;
        } else {
            return [UIApplication sharedApplication].statusBarFrame.size.height;
        }
    }
}

- (CGFloat)fwNavigationBarHeight
{
    if (self.navigationController.navigationBarHidden) {
        return 0.0;
    } else {
        return self.navigationController.navigationBar.frame.size.height;
    }
}

- (CGFloat)fwTabBarHeight
{
    if (self.tabBarController.tabBar.hidden) {
        return 0.0;
    } else {
        return self.tabBarController.tabBar.frame.size.height;
    }
}

@end

File: FWFramework/Application/App/Toolkit/UIKit/UIScreen+FWFramework.swift
//
//  UIScreen+FWFramework.swift
//  FWFramework
//
//  Created by wuyong on 2019/6/28.
//  Copyright © 2019 wuyong.site. All rights reserved.
//

import Foundation

/// 屏幕尺寸
public var FWScreenSize: CGSize {
    return UIScreen.main.bounds.size
}

/// 屏幕宽度
public var FWScreenWidth: CGFloat {
    return UIScreen.main.bounds.size.width
}

/// 屏幕高度
public var FWScreenHeight: CGFloat {
    return UIScreen.main.bounds.size.height
}

/// 屏幕像素比例
public var FWScreenScale: CGFloat {
    return UIScreen.main.scale
}

/// 屏幕分辨率
public var FWScreenResolution: CGSize {
    return CGSize(width: UIScreen.main.bounds.size.width * UIScreen.main.scale, height: UIScreen.main.bounds.size.height * UIScreen.main.scale)
}

/// 判断屏幕尺寸
///
/// - Parameters:
///   - width: 屏幕宽度
///   - height: 屏幕高度
/// - Returns: 是否为该尺寸
public func FWIsScreenSize(_ width: CGFloat, _ height: CGFloat) -> Bool {
    return UIScreen.main.bounds.size.equalTo(CGSize(width: width, height: height))
}

/// 判断屏幕分辨率
///
/// - Parameters:
///   - width: 分辨率宽度
///   - height: 分辨率高度
/// - Returns: 是否为该分辨率
public func FWIsScreenResolution(_ width: CGFloat, _ height: CGFloat) -> Bool {
    return FWScreenResolution.equalTo(CGSize(width: width, height: height))
}

/// 是否是3.5英寸屏幕
public var FWIsScreen35: Bool {
    return FWIsScreenSize(320, 480)
}

/// 是否是4.0英寸屏幕
public var FWIsScreen40: Bool {
    return FWIsScreenSize(320, 568)
}

/// 是否是4.7英寸屏幕
public var FWIsScreen47: Bool {
    return FWIsScreenSize(375, 667)
}

/// 是否是5.5英寸屏幕
public var FWIsScreen55: Bool {
    return FWIsScreenSize(414, 736)
}

/// 是否是5.8英寸屏幕
public var FWIsScreen58: Bool {
    return FWIsScreenSize(375, 812)
}

/// 是否是6.1英寸屏幕
public var FWIsScreen61: Bool {
    return FWIsScreenResolution(828, 1792)
}

/// 是否是6.5英寸屏幕
public var FWIsScreen65: Bool {
    return FWIsScreenResolution(1242, 2688)
}

/// 是否是iPhoneX系列全面屏幕
public var FWIsScreenX: Bool {
    return FWIsScreenSize(375, 812) || FWIsScreenSize(414, 896)
}

/// 状态栏高度
public var FWStatusBarHeight: CGFloat {
    return FWIsScreenX ? 44.0 : 20.0
}

/// 导航栏高度
public let FWNavigationBarHeight: CGFloat = 44.0

/// 标签栏高度
public var FWTabBarHeight: CGFloat {
    return FWIsScreenX ? 83.0 : 49.0
}

/// 顶部栏高度，包含状态栏、导航栏
public var FWTopBarHeight: CGFloat {
    return FWStatusBarHeight + FWNavigationBarHeight;
}

/// 底部栏高度，包含标签栏
public var FWBottomBarHeight: CGFloat {
    return FWTabBarHeight
}
